library/std/src/sys/uefi/stdio.rs

```diff
@@ -4,45 +4,48 @@ use crate::mem::MaybeUninit;
 use crate::os::uefi;
 use crate::ptr::NonNull;
 
-const MAX_BUFFER_SIZE: usize = 8192;
+pub struct Stdin {
+    pending: Option<char>,
+}
 
-pub struct Stdin;
 pub struct Stdout;
 pub struct Stderr;
 
 impl Stdin {
     pub const fn new() -> Stdin {
-        Stdin
+        Stdin { pending: None }
     }
 }
 
 impl io::Read for Stdin {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
         let st: NonNull<r_efi::efi::SystemTable> = uefi::env::system_table().cast();
         let stdin = unsafe { (*st.as_ptr()).con_in };
 
-        // Try reading any pending data
-        let inp = match read_key_stroke(stdin) {
-            Ok(x) => x,
-            Err(e) if e == r_efi::efi::Status::NOT_READY => {
-                // Wait for keypress for new data
-                wait_stdin(stdin)?;
-                read_key_stroke(stdin).map_err(|x| io::Error::from_raw_os_error(x.as_usize()))?
+        // Write any pending character
+        if let Some(ch) = self.pending {
+            if ch.len_utf8() > buf.len() {
+                return Ok(0);
             }
-            Err(e) => {
-                return Err(io::Error::from_raw_os_error(e.as_usize()));
-            }
-        };
+            ch.encode_utf8(buf);
+            buf = &mut buf[ch.len_utf8()..];
+            self.pending = None;
+        }
+
+        // Try reading any pending data
+        let inp = simple_text_input_read(stdin)?;
 
         // Check if the key is printiable character
-        if inp.scan_code != 0x00 {
+        if inp == 0x00 {
```

Comment:
Hmm. This is not actually sufficient to check if it's printable, as e.g. 0x02 and 0x03 are not printable. What is the _actual_ criteria we're applying? I assume that if data pours in over stdin, we actually just want to pass it through, as long as it's validly encoded, even if the output would be nonprintable on the other side? Why _don't_ we pass null bytes through? And if we want to reject nonprintable inputs, why _would_ we reject e.g. an input of 0x7 (aka BEL aka Bell)?

Reply:
Pretend for the sake of this discussion that I have a magic keyboard that can issue arbitrary scancodes, btw, if you're having trouble imagining when this would happen.

Because I do.
Like, actually.

Reply:
So the UEFI spec says this: " The UnicodeChar is the actual printable character or is zero if the key does not represent a printable character (control key, function key, etc.)."
https://uefi.org/sites/default/files/resources/UEFI_Spec_2_10_Aug29.pdf

Here I have changed it check the `unicode_char`, (`simple_text_input_read` returns unicode_char)

Reply:
Ahh I see! Thank you. For some reason my initial assumption was they'd just copy things through instead of doing validation.